plonky3/src/prover.rs

```diff
@@ -1,71 +1,470 @@
+                    key.preprocessed.get(name).map(|preprocessed| {
+                        (
+                            // pick the preprocessed data for this table in the correct size
+                            &preprocessed[&(1 << table.log_degree())].prover_data,
+                            vec![vec![
+                                zeta,
+                                table.trace_domain(state.pcs).next_point(zeta).unwrap(),
+                            ]],
+                        )
+                    })
+                })
+            })
+            .into_iter()
+            .flatten();
+
+        let trace_data_and_points_per_stage: Vec<(_, Vec<Vec<_>>)> = state
+            .processed_stages
+            .iter()
+            .map(|processed_stage| {
+                let points = self
+                    .tables
+                    .values()
+                    .map(|input| {
+                        vec![
+                            zeta,
+                            input.trace_domain(state.pcs).next_point(zeta).unwrap(),
+                        ]
+                    })
+                    .collect();
+                (&processed_stage.prover_data, points)
+            })
+            .collect();
+
+        let quotient_opening_points: Vec<_> = (0..self.quotient_chunks_count())
+            .map(|_| vec![zeta])
+            .collect();
+
+        let (opened_values, proof) = state.pcs.open(
+            preprocessed_data_and_opening_points
+                .chain(trace_data_and_points_per_stage)
+                .chain(once((&quotient_data, quotient_opening_points)))
+                .collect(),
+            state.challenger,
+        );
+
+        let mut opened_values = opened_values.into_iter();
+
+        // maybe get values for the preprocessed columns
+        let preprocessed: Vec<_> = if let Some(proving_key) = proving_key {
+            state
+                .program
+                .tables
+                .keys()
+                .map(|name| {
+                    proving_key.preprocessed.contains_key(name).then(|| {
+                        let value = opened_values.next().unwrap();
+                        assert_eq!(value.len(), 1);
+                        StageOpenedValues {
+                            local: value[0][0].clone(),
+                            next: value[0][1].clone(),
+                        }
+                    })
+                })
+                .collect()
+        } else {
+            vec![None; state.program.table_count()]
+        };
+
+        // get values for the traces
+        let traces_by_table_by_stage: Vec<Vec<_>> = state.processed_stages.iter().fold(
+            vec![vec![]; state.program.table_count()],
+            |mut traces_by_table, _| {
+                let mut values = opened_values.next().unwrap();
+                for (values, v) in traces_by_table.iter_mut().zip_eq(values.iter_mut()) {
+                    assert_eq!(v.len(), 2);
```

Comment:
Good question! The output of `pcs.open` returns an `OpenedValues<F>` defined as:
```
pub type OpenedValues<F> = Vec<OpenedValuesForRound<F>>;
pub type OpenedValuesForRound<F> = Vec<OpenedValuesForMatrix<F>>;
pub type OpenedValuesForMatrix<F> = Vec<OpenedValuesForPoint<F>>;
pub type OpenedValuesForPoint<F> = Vec<F>;
```

Here `values` is an `OpenValuesForRound<F>` obtained as a element of `opened_values`.
`v` is an element of that, so it's a `OpenedValuesForMatrix<F>`.
Now, it has size two, because for the rounds corresponding to the traces (or matrices!), we open the commitment at two points: `zeta` and `zeta_next`.
All of this is defined in the inputs to `pcs.open`, in this case [here](https://github.com/powdr-labs/powdr/blob/40e385ce8678b7825bf606995aad371a50358ebd/plonky3/src/prover.rs#L156-L160)